`Development`: Add rules for pull request titles into the documentation (#4259)
ls1intum/Artemis: docs/dev/guidelines/development-process.rst

******************************************
Development Process (How to Merge Changes)
******************************************

.. contents:: Content of this document
    :local:
    :depth: 2

Naming Conventions for GitHub Pull Requests
===========================================

> 1. The first term is a main feature of Artemis and is using code highlighting, e.g.  “``Programming exercises``:”

Reviewer: ```suggestion
1. The first term is a main feature of Artemis and is using code highlighting, e.g.  “``Programming exercises``:”.
```

    1. Possible feature tags are: ``Programming exercises``, ``Modeling exercises``, ``Text exercises``, ``Quiz exercises``, ``File upload exercises``, ``Lectures``, ``Exam mode``, ``Assessments``, ``Discussion``, ``Notifications``. More tags are possible if they make sense.
    2. If no feature makes sense, and it is a pure development or test improvement, we use the term “``Development``:”. More tags are also possible if they make sense.
    3. Everything else belongs to the ``General`` category.

> 2. The colon is not highlighted

Reviewer: ```suggestion
2. The colon is not highlighted.
```

3. After the colon, there should be a verbal form that is understandable by end users and non-technical persons, because this will automatically become part of the release notes.

    1. The text should be short, non-capitalized (except the first word) and should include the most important keywords. Do not repeat the feature if it is possible.

>     2. We generally distinguish between Bugfixes (the verb “Fix”) and Improvements (all kinds of verbs) in the release notes. This should be immediately clear from the title.

Reviewer: ```suggestion
    2. We generally distinguish between bugfixes (the verb “Fix”) and improvements (all kinds of verbs) in the release notes. This should be immediately clear from the title.
```

    3. Good examples:

        - “Allow instructors to delete submissions in the participation detail view”
        - “Fix an issue when clicking on the start exercise button”
        - “Add the possibility for instructors to define submission policies”

Steps to Create and Merge a Pull Request
========================================

0. Precondition -> only Developer
---------------------------------

> * Limit yourself to one functionality per pull request
> * Split up your task in multiple branches & pull requests if necessary
> * `Commit Early, Commit Often, Perfect Later, Publish Once <https://speakerdeck.com/lemiorhan/10-git-anti-patterns-you-should-be-aware-of>`_

Reviewer: ```suggestion
* Limit yourself to one functionality per pull request.
* Split up your task in multiple branches & pull requests if necessary.
* `Commit Early, Commit Often, Perfect Later, Publish Once. <https://speakerdeck.com/lemiorhan/10-git-anti-patterns-you-should-be-aware-of>`_
```

1. Start Implementation -> only Developer
-----------------------------------------

> * `Open draft pull request. <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request>`_ Allows for code related questions and discussions

Reviewer: ```suggestion
* `Open a draft pull request. <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request>`_ This allows for code related questions and discussions.
```

2. Implementation is "done" -> only Developer
---------------------------------------------

> * Make sure all steps in the `Checklist <https://github.com/ls1intum/Artemis/blob/develop/.github/PULL_REQUEST_TEMPLATE.md>`_ are completed
> * Add or update the "Steps for Testing" in the description of your pull request
> * Make sure that the changes in the pull request are only the ones necessary
> * Mark the pull request as `ready for review <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request>`_

Reviewer: ```suggestion
* Make sure all steps in the `Checklist <https://github.com/ls1intum/Artemis/blob/develop/.github/PULL_REQUEST_TEMPLATE.md>`_ are completed.
* Add or update the "Steps for Testing" in the description of your pull request.
* Make sure that the changes in the pull request are only the ones necessary.
* Mark the pull request as `ready for review. <https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/changing-the-stage-of-a-pull-request>`_
```

3. Review
---------

> Developer
> ---------

Reviewer: ```suggestion
Developer
^^^^^^^^^
```
This should be a sub-sub-section.

* Organize or join a testing session. Especially for large pull requests this makes testing a lot easier.
* Actively look for reviews. Do not just open the pull request and wait.

> Reviewer
> ---------

Reviewer: ```suggestion
Reviewer
^^^^^^^^
```

* Perform the "Steps for Testing" and verify that the new functionality is working as expected.
* Verify that related functionality is still working as expected.

> * Check the Changes to

Reviewer: ```suggestion
* Check the changes to
```

>     * Conform with the code style
>     * Make sure you can easily understand the code
>     * Make sure that (extensive) comments are present where deemed necessary
>     * Performance is reasonable (e.g. Number of database queries or HTTP calls)

Reviewer: ```suggestion
    * conform with the code style.
    * make sure you can easily understand the code.
    * make sure that (extensive) comments are present where deemed necessary.
    * performance is reasonable (e.g. number of database queries or HTTP calls).
```
The continuation of a sentence should start with a lowercase letter.

* Submit your comments and status ((thumbs up) Approve or (thumbs down) Request Changes) using GitHub.

>     * Explain what you did (test, review code) and on which test server in the review comment

Reviewer: ```suggestion
    * Explain what you did (test, review code) and on which test server in the review comment.
```

4. Respond to review
--------------------

> Developer
> ---------

Reviewer: ```suggestion
Developer
^^^^^^^^^
```

> * Use the pull request to discuss comments or ask questions
> * Update your code where necessary
> * Revert to Draft if the changes will take a while during which review is not needed/possible
> * Set back to ready for review afterwards
> * Notify the reviewer(s) once your revised version is ready for the next review
> * Comment on "inline comments" (e.g. "Done")

Reviewer: ```suggestion
* Use the pull request to discuss comments or ask questions.
* Update your code where necessary.
* Revert to draft if the changes will take a while during which review is not needed/possible.
* Set back to ready for review afterwards.
* Notify the reviewer(s) once your revised version is ready for the next review.
* Comment on "inline comments" (e.g. "Done").
```

> Reviewer
> ---------

Reviewer: ```suggestion
Reviewer
^^^^^^^^
```

> * Respond to questions raised by the reviewer
> * Mark Conversations as Resolved if the change is sufficient

Reviewer: ```suggestion
* Respond to questions raised by the reviewer.
* Mark conversations as resolved if the change is sufficient.
```

> Iterate steps 3 & 4 until ready for merge (- all reviewers approve (thumbs up) )
> --------------------------------------------------------------------------------

Reviewer: ```suggestion
Iterate steps 3 & 4 until ready for merge (- all reviewers approve (thumbs up) )
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

5. Merge
--------

> A Project Maintainer merges your changes to develop

Reviewer: ```suggestion
A project maintainer merges your changes into the ``develop`` branch.
```
